finite_state_machine/state_machine.py

```diff
@@ -72,12 +47,37 @@ def transition(source, target, conditions=None, on_error=None):
             raise ValueError("on_error needs to be a bool, int or string")
 
     def transition_decorator(func):
-        func.__fsm = TransitionMeta(func.__name__)
-        if isinstance(source, (list, tuple, set)):
-            for item in source:
-                func.__fsm.add_transition(item, target, on_error, conditions)
+        mems = inspect.getmembers(func)
+        state_machine_instance = [
+            mem[1]["StateMachine"] for mem in mems if mem[0] == "__globals__"
+        ][0]
+        func.__fsm = Transition(
+            name=func.__name__,
+            source=source,
+            target=target,
+            conditions=conditions,
+            on_error=on_error,
+        )
+        # need to optimize
+        if hasattr(state_machine_instance, "__fsm"):
+            if isinstance(source, list):
+                for src in source:
+                    if src in state_machine_instance.__fsm:
+                        state_machine_instance.__fsm[src].append(target)
+                    else:
+                        state_machine_instance.__fsm[src] = [target]
+            else:
+                if source in state_machine_instance.__fsm:
+                    state_machine_instance.__fsm[src].append(target)
+                else:
+                    state_machine_instance.__fsm[src] = [target]
         else:
-            func.__fsm.add_transition(source, target, on_error, conditions)
+            if isinstance(source, list):
+                state_machine_instance.__fsm = {}
+                for src in source:
+                    state_machine_instance.__fsm[src] = [target]
+            else:
+                state_machine_instance.__fsm.source = [target]
```

Comment:
This probably needs to be optimized. Checks if the StateMachine has an attribute "__fsm", and appends a target to an existing source or creates an entry in a dict of structure {source: [target]}. 

Two questions:
1. Any advice on how to optimize this?
2. I like collections.defaultdict as the data type. Should I build that out?

Reply:
`collections.defaultdict` would definitely reduce one level of nesting. I would try that out and see what you are left with

Will be easier to find the next optimization

Reply:
Since our types at the moment are booleans, ints, and strings. We can probably start with the default type of a set. Not sure how that would work for enums.

I still need to investigate that.